src/algorithms/digi/TOFPulseGeneration.cc

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Souvik Paul, Chun Yuen Tsang, Prithwish Tribedy
+// Special Acknowledgement: Kolja Kauder
+//
+// Convert energy deposition into ADC pulses
+// ADC pulses are assumed to follow the shape of landau function
+
+#include <DDRec/CellIDPositionConverter.h>
+#include <RtypesCore.h>
+#include <algorithm>
+#include <gsl/pointers>
+#include <unordered_map>
+#include <vector>
+
+#include "TMath.h"
+#include "TOFPulseGeneration.h"
+#include "algorithms/digi/TOFHitDigiConfig.h"
+
+
+namespace eicrecon {
+
+double TOFPulseGeneration::_Landau(double x, double mean, double std) const {
+  double C = -113.755;
+  return C * TMath::Landau(x, mean, std, kTRUE);
+}
+
+void TOFPulseGeneration::process(const TOFPulseGeneration::Input& input,
+                                 const TOFPulseGeneration::Output& output) const {
+  const auto [simhits] = input;
+  auto [rawADCs] = output;
+
+  double Vm = m_cfg.Vm;
+  double tMin = m_cfg.tMin;
+  double tMax = m_cfg.tMax;
+  int adc_range = m_cfg.adc_range;
+  int tdc_range = m_cfg.tdc_range;
+  int nBins = m_cfg.tdc_range;
+
+  // signal sum
+  // NOTE: we take the cellID of the most energetic hit in this group so it is a real cellID from an
+  // MC hit
+  std::unordered_map<dd4hep::rec::CellID, std::vector<double>> adc_sum;
+  double interval = (tMax - tMin) / (nBins - 1);
+
+  for (const auto& hit : *simhits) {
+    auto cellID          = hit.getCellID();
+    double sum_charge = 0.0;
+    double mpv_analog = 0.0;
+
+    double  time       = hit.getTime();
+    double  charge     = hit.getEDep();
+    // reduce computation power by not simulating low-charge hits
+    if(charge < m_cfg.ignore_thres) continue;
+
+    auto& ADCs = adc_sum[cellID];
+    if(ADCs.size() == 0) ADCs.resize(nBins, 0);
+
+    mpv_analog = time + m_cfg.risetime;
```

Comment:
I think I understand at least some of your concern. If the hit time lies outside of `tMin `and `tMax`, the `landau_min `will not reflects the peak height. I can replace `landau_min `with `double landau_min = this -> _Landau(mvp_analog, mpv_analog, m_cfg.sigma_analog);`

However, I am not sure if I understand the last part about `tMin `rounding down to 25 ps. Since there is no clock in the current implementation, TOF time is perfectly aligned with the collision time, i.e. t=0 is when collision occur and TOF pulse begins. (I realized that `tMin = 0.1 `in `TOFHitDigiConfig`, but that is a mistake. It should have been 0). Do you want me to change that code such that `tMin `is a variable instead in preparation for a clock class in the future where the beginning time is taken as an input? I can do that, but if I misunderstood your comments, please let me know. 

I don't think we should set `tMax=time_signal_returns_below_threshold` and just save the entire pulse until the stop signal (in my case, it's always 25 ns). If we want to discard pulses below threshold, we can do it downstream in other classes.

Reply:
The problem is that the collision time is not always at `t=0`, within a bunch crossing it can vary between around +/-100 ps from where the centre of the bunches cross z=0. Ontop of that the simulation needs to deal with collisions from offset bunches such that the time associated with the hit from the simulation should be able to be treated reasonably equally with any value.

Reply:
Let see if I understand correctly. `hit.getTime()` returns Geant time, which always start at zero when collision occur. But like you said, bunch crossing can vary around +/- 100 ps. To account for this, instead of offsetting the Geant time, you want me to offset the TOF pulse timing windows, is that correct?

I think similar arguments extend to offset bunches. Let say a collision occurs at 10ns after the beginning of a 25ns cycle, i.e. detector only save the final 15ns to disk, wouldn't it make more sense to add 10ns to `hit.getTime()` rather than to shift the pulse generation windows by 10ns?

Reply:
What is incorrect here is that the Geant/simulation time starts at 0 at the start of a simulation. The MonteCarlo event samples used in the simulation already have times attributed to them which is meant to replicate the distribution of collisions throughout time. All of the times recorded in the hits already has this offset included so your algorithm has to be able to account for an arbitrary time coming from BTOF hits.

Reply:
I understand now. In that case, can you tell me how to fetch the event start time please? I can replace `double time = hit.getTime();` with `double time = hit.getTime() - eventStartTime;`.

Reply:
This also isn't quite what I think you want, I will try and explain.

Using your Landau model, the analogue pulse will have an identical distribution starting from the simulation hit time. This means you can just use the `mpv_analog = m_cfg.risetime` to give the pulse shape.

When you are creating the `time_series` you are explicitly adding some digitization to your analogue pulse. The edges of the steps in the `time_series` for different hits, ideally should all line up so there is no added ambiguity when summing pulses offset in time, or bias introduced by always aligning the simulation hit with and edge of a time step.

In order to achieve this, the `tMin` value and `time` of the `time_series` should be adjusted to align with integer multiples of the `interval` of the `time_series` for each hit individually. Something equivalent to e.g.
`startTime = (int)(time/interval)*interval`
`tMin = time - startTime - interval`
`time_series.setTime(startTime)`

It might be easier to leave out summing of the `time_series` initially, or move it to a separate algorithm, but this will require checking all series associated with a pixel and seeing if they overlap based on their `startTime` and length*`interval`. The ADCCounts is just an array so doesn't have to be the same length for every time series. Using the offset in units of interval from the series start time the adc values can be added together, creating a longer vector than either original pulse.

Reply:
I am still a bit confused. Perhaps it will be better if you can clarify the following question: In epic simulation, what does t=0 mean? At first, I thought t=0 means the collision happens, but then you told me Geant doesn't always start at t=0. Therefore, I guess t=0 corresponds to the time when 40 MHz clock begins a new cycle? 

If that is wrong, then please let me know the physical significance of t=0 in the simulation. If t=0 indeed corresponds to the beginning of a 40MHz cycle, then by setting `tMin = 0` and `mpv_analog = time + m_cfg.risetime`, I should have aligned all of my hits with respect to the start time, shouldn't I? The time bin edges will be t = [0, interval, 2*interval, 3*interval,...] for all hits in the same events.

Reply:
I have updated my code such that the `tMin `is integer multiple of `interval`. Please take a look at [TOFPulseGeneration.cc](https://github.com/eic/EICrecon/blob/pr/BTOF-digitization-clusterization/src/algorithms/digi/TOFPulseGeneration.cc) and see if that's what you are looking for.

Reply:
This still looks like it is coving a static range of times set in the config, `tMin `and `tMax `need to change for each hit. Inside the loop you could use 
`tMin = _DigitizeTime(time)`
`tMax = tMin+m_cfg.tMax`

What is the `+1e-3` doing in your DigitizeTime function?

Given that you cannot rely on multiple hits being within the same time range or their start time aligning, the way you handle the `adc_sum` will either have to be moved to another algorithm or made more complicated here.

Reply:
> If a hit arrives at time > 25 ns, it will just be discarded.
This is the key point - you cannot discard these hits.

The algorithm needs to work for hits arbitrarily distributed through time.

Reply:
It's that how LGAD works in EPIC? I thought the end of a 40 MHz cycle marks the end of an event and everything that happens after that are discarded? There are only 1024 TDC bins, how can it save anything beyond 25 ns?

Reply:
I don't know the details of the EICROC/AC-LGAD myself but the 1024 TDC bins will be the finest time binning of the electronics over the 25ns. It might be that the EICROC only outputs this 10 bit tdc but there will also be another layer(s) of electronics which will tag the 25ns bin which the hit occurred in so that a global time can be reconstructed.

Unfortunately I don't know of any specific links to help you.

Reply:
It's quite unfortunate because we've asked multiple people here at BNL and none of them claim to know how clocks work. My advisor, Prithwish Tribedy, seems to be the one with the most intimate knowledge on these kind of thing, but he is on vacation until after thanksgiving.

I implemented what I believe Prithwish thinks will happen to EPIC, i.e. recording starts at the beginning of 40 MHz clock cycle and ends when that cycle ends. After that, there will be 143 ns of dead time. My TOFPulseGenerator reflects my understanding of how timing works. 

From a physics point of view, if an event starts at t = 0 (which may not be because of mismatch between the 100 MHz experiment clock and the 40 MHz EIC clock), a hit after 25 ns isn't really possible for BTOF. Straight line distance between the vertex and the furthest edge of BTOF = sqrt(1.5m^2 + 0.6m^2) =  1.62m, so a hit at 25 ns implies a speed of 1.6/25e-9 = 0.22c. If this is a proton, it would have a momentum of just ~200 MeV/c. This is very slow considering our CM beam energy of ~100 GeV.